httpclient/httpclient.go

```diff
@@ -0,0 +1,41 @@
+package httpclient
+
+import (
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// GetURL retrieves data from an URL.
+// It uses some technique to slow down the requests if it get a 429 (Too Many Requests) response.
+func GetURL(URL string) ([]byte, error) {
+	var sleep time.Duration
+	const timeout = time.Duration(10 * time.Second)
+
+	client := http.Client{
+		Timeout: timeout,
+	}
+
+	for {
+		resp, err := client.Get(URL)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode == 429 {
+			sleep = sleep + (5 * time.Minute)
+			log.Infof("Rate limit reached, sleep %v minutes\n", sleep)
```

Comment:
Io ho usato https://godoc.org/golang.org/x/time/rate per fare rate limiting. Se strutturi il codice per avere una libreria "bitbucket", per esempio, che fa tutte le chiamate a bitbucket, puoi usare un'istanza di Limiter su tutti i metodi per imporre un rate limiting globale di tot chiamte/secondo, indipendentemente da quanta concorrenza hai (magari hai parti di codice diverse che fanno in parallelo chiamate ad endpoint diversi di Bitbucket, ma globalmente imposti un limit di 5 req/s per dire).